csrc/optimization/alias_analysis.cpp

```diff
@@ -0,0 +1,79 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <unordered_map>
+#include <vector>
+
+#include <ir/interface_nodes.h>
+#include <ir/internal_base_nodes.h>
+#include <ir/utils.h>
+#include <optimization/alias_analysis.h>
+
+namespace nvfuser::optimization {
+
+namespace {
+
+// Returns whether the input TensorView is contiguous on every non-broadcast
+// IterDomain.
+bool isContiguous(const TensorView& tv) {
+  NVF_ERROR(tv.nDims() == tv.getContiguity().size());
+  for (size_t i = 0; i < tv.nDims(); i++) {
+    if (!tv.axis(i)->isBroadcast() && !tv.getContiguity()[i]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+// Finds aliases of `source` and stores the findings in `alias_to_source`.
+void findAliasesOfSource(
+    const TensorView* source,
+    AliasAnalysisResult& alias_to_source) {
+  // The current implementation does the bare minimum to detect some aliasing
+  // that the codegen can use to generate a kernel skipping unnecessary
+  // computation.
+  std::queue<const TensorView*> q;
+  if (!source->hasAllocation() && isContiguous(*source)) {
+    q.push(source);
+  }
+
+  while (!q.empty()) {
+    const TensorView* in_tv = q.front();
+    q.pop();
+
+    for (Expr* use : in_tv->uses()) {
+      if (!use->isA<ViewOp>()) {
+        continue;
+      }
+
+      Val* out = use->output(0);
+      TensorView* out_tv = dynamic_cast<TensorView*>(out);
+      if (out_tv == nullptr) {
+        continue;
+      }
+
+      if (!out_tv->hasAllocation() && isContiguous(*out_tv)) {
+        q.push(out_tv);
+        alias_to_source[out_tv] = in_tv;
+      }
```

Comment:
Can you comment more here on why this guarantees that this `ViewOp` can be accomplished without a copy? I had gone through this analysis a few months ago and I thought you needed to propagate contiguity from root to rfactor to do the check. Then when you encounter a `Merge` where the outer of the two merged `IterDomain`s is discontiguous you know that you need a copy since a single stride can no longer represent that transformation.

Reply:
I'm pretty sure this will show my ignorance of various domains, but let me think aloud 😄. I think the check here is sufficient (but not necessary). Every TensorView added to the queue has a contiguous leaf domain. So a ViewOp entering the if-then must have a contiguous root and leaf domain and the default major-to-minor memory layout. I think this guarantees the same storage can be reused. Can you think of a counter example?

Reply:
I think if the input is contiguous in the normal stride order then any reshape is possible without a copy. However consider an irregular stride order like channels last, but still contiguous. Reshaping that to 1D is impossible without a copy.

Reply:
I agree that your check is sufficient but not necessary. Thanks for explaining.

Reply:
Where is stride order checked?

Reply:
> I think if the input is contiguous in the normal stride order then any reshape is possible without a copy.

Good point -- we can always treat `t` as `f` without breaking functionality. I'll try to change my code reflecting that. 

> Where is stride order checked?

AFAICT, we don't keep stride order in `TensorView`. Instead, we convert stride order to allocation domain in `TensorView`: https://github.com/NVIDIA/Fuser/blob/cc9b43843c67394f02ee7a158105e762942c5e1a/csrc/python_frontend/fusion_record.h#L1351. My code does check that the allocation domain is empty, meaning the default major-to-minor layout.

Reply:
> Good point -- we can always treat t as f without breaking functionality. I'll try to change my code reflecting that.

I don't think it's true any more. The frontend can request a non-contiguous stride order for a fusion output and expect the reshape that produces the output to copy data from contiguous to non-contiguous. While nvFuser wants to choose an allocation domain that preserves aliasing when possible, it shouldn't conflict with what's requested by the frontend.